src/libostree/ostree-sign-ed25519.c

```diff
@@ -0,0 +1,588 @@
+ * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
+ * Boston, MA 02111-1307, USA.
+ *
+ * Authors:
+ *  - Denis Pynkin (d4s) <denis.pynkin@collabora.com>
+ */
+
+#include "config.h"
+
+#include "ostree-sign-ed25519.h"
+#ifdef HAVE_LIBSODIUM
+#include <sodium.h>
+#endif
+
+#undef G_LOG_DOMAIN
+#define G_LOG_DOMAIN "OSTreeSign"
+
+#define OSTREE_SIGN_ED25519_NAME "ed25519"
+
+#define OSTREE_SIGN_METADATA_ED25519_KEY "ostree.sign.ed25519"
+#define OSTREE_SIGN_METADATA_ED25519_TYPE "aay"
+
+#if 0
+#define SIGNIFY_COMMENT_HEADER "untrusted comment:"
+#define SIGNIFY_ID_LENGTH 8
+#define SIGNIFY_MAGIC_ED25519 "Ed"
+#endif
+
+struct _OstreeSignEd25519
+{
+  GObject parent;
+  gboolean initialized;
+  guchar *secret_key;
+  GList *public_keys;
+};
+
+static void
+ostree_sign_ed25519_iface_init (OstreeSignInterface *self);
+
+G_DEFINE_TYPE_WITH_CODE (OstreeSignEd25519, ostree_sign_ed25519, G_TYPE_OBJECT,
+        G_IMPLEMENT_INTERFACE (OSTREE_TYPE_SIGN, ostree_sign_ed25519_iface_init));
+
+static void
+ostree_sign_ed25519_iface_init (OstreeSignInterface *self)
+{
+  g_debug ("%s enter", __FUNCTION__);
+
+  self->data = ostree_sign_ed25519_data;
+  self->get_name = ostree_sign_ed25519_get_name;
+  self->metadata_key = ostree_sign_ed25519_metadata_key;
+  self->metadata_format = ostree_sign_ed25519_metadata_format;
+  self->metadata_verify = ostree_sign_ed25519_metadata_verify;
+  self->set_sk = ostree_sign_ed25519_set_sk;
+  self->set_pk = ostree_sign_ed25519_set_pk;
+  self->add_pk = ostree_sign_ed25519_add_pk;
+  self->load_pk = ostree_sign_ed25519_load_pk;
+}
+
+static void
+ostree_sign_ed25519_class_init (OstreeSignEd25519Class *self)
+{
+  g_debug ("%s enter", __FUNCTION__);
+}
+
+static void
+ostree_sign_ed25519_init (OstreeSignEd25519 *self)
+{
+  g_debug ("%s enter", __FUNCTION__);
+
+  self->initialized = TRUE;
+  self->secret_key = NULL;
+  self->public_keys = NULL;
+
+#ifdef HAVE_LIBSODIUM
+  if (sodium_init() < 0) 
+    {
+      self->initialized = FALSE;
+      g_warning ("libsodium library couldn't be initialized");
+    }
+#else
+  g_error ("ed25519 signature isn't supported");
+#endif /* HAVE_LIBSODIUM */
+}
+
+gboolean ostree_sign_ed25519_data (OstreeSign *self,
+                                   GBytes *data,
+                                   GBytes **signature,
+                                   GCancellable *cancellable,
+                                   GError **error)
+{
+
+  g_debug ("%s enter", __FUNCTION__);
+  g_return_val_if_fail (OSTREE_IS_SIGN (self), FALSE);
+  OstreeSignEd25519 *sign = ostree_sign_ed25519_get_instance_private(OSTREE_SIGN_ED25519(self));
+
+#ifdef HAVE_LIBSODIUM
+  g_autofree guchar *sig = NULL;
+#endif
+
+  if ((sign->initialized != TRUE) || (sign->secret_key == NULL))
+    {
+      g_set_error_literal (error, G_IO_ERROR, G_IO_ERROR_FAILED,
+                          "Not able to sign: libsodium library isn't initialized properly");
+      goto err;
+    }
+#ifdef HAVE_LIBSODIUM
```

Comment:
Maybe this whole object should not be built if libsodium is not available.
Otherwise the code is cluttered with `#ifdef HAVE_LIBSODIUM`. In the long term, you'd need to build the code twice if you wants to be sure it builds.

Reply:
We do have CI that builds in multiple configurations at least.

Reply:
correct, I think it should be the separate MR, since the current one is too huge already